### DISCUSSION TITLE
Launching gui_based_control vs. launch_ctrk_interface? Or are these not mutually exclusive?

### DISCUSSION THREAD

**Discussion by sunshineinsandiego** [surgical-robotics-ai/surgical_robotics_challenge#78] [Q&A]
Hi - Following the instructions [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/tree/master/scripts) to try to launch GUI control of the PSM arms / ECM, but not understanding how exactly to do this.

(1) It is written:

"Simply run the `launch_crtk_interface.py` script. This script will consequently use the `psm_arm.py`, `ecm_arm.py` and `scene.py` scripts and create ROS topics to publish and receive commands."

-- My belief is that I should run `$ python3 launch_crtk_interface.py` in a 3rd terminal with my other two terminals running `$ roscore` and `$ ./ambf_simulator --launch_file <surgical_robotics_challenge>/launch.yaml -l 0,1,3,4,14,15 -p 120 -t 1 --override_max_comm_freq 120`.

(2) Then I should run `$ gui_based_control.py` in a 4th terminal to access the GUIs that will interface with the ROS topics launched by the `launch_crtk_interface.py` script?

OR 

(3) Are these two separate concepts? The `launch_crtk_interface.py` should be used with ROS messages while the `gui_based_control.py` already includes the interface with the simulated PSMs/ECM? In which case `gui_based_control.py` should be run with the other two processes being `roscore` and `./ambf_simulator`?

Thanks!

**Comment by adnanmunawar** [comment]
Hello,

Thanks for the detailed question. Let me clarify. 

1. The objects (PSMs and ECM) in the surgical robotics challenge (SRC) assets can be controlled using two interfaces: **1) Method based API** and **2) ROS Topics based API**. The **ROS Topics based API** is implemented by the `launch_crtk_interface.py` script which internally uses the methods based API but to the user it provides ROS topics which relays the state of the PSMs, ECM and the scene, and complementary topics that allow sending commands to the PSMs, ECM and the Scene (i.e. Needle etc).
2. The `gui_based_control.py` script uses the **Method based API** and thus does not need the `launch_crtk_interface.py` script to be running. This is what your point `3` implies and that is correct. You should note that the `gui_based_control.py` is only an example script which allows controlling the PSMs and ECM using GUI sliders. This is not intuitive enough to allow one to complete the suturing. 
3. You can look at the two examples [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/tree/master/scripts/surgical_robotics_challenge/examples) namely [`interface_via_method_api.py`](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/scripts/surgical_robotics_challenge/examples/interface_via_method_api.py) and [`interface_via_crtk_ros_api.py`](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/scripts/surgical_robotics_challenge/examples/interface_via_crtk_ros_api.py) to see how **Method Based API** and **ROS Topics based API** interfaces are used. NOTE: For both the examples, you will need `roscore` and the `./ambf_simulator --launch_file <surgical_robotics_challenge>/launch.yaml -l 0,1,3,4,14,15 -p 120 -t 1 --override_max_comm_freq 120` processes to be running. Also, for running `interface_via_crtk_ros_api.py`, you will also need to run the `launch_crtk_interface.py` beforehand.

**Reply by sunshineinsandiego** [reply]
Sorry for the delayed response! Thanks, this was super helpful. I think I understand this now.
